# Diabetes-Onset Detection
### Created by Harshit Agarwal @ 20-10-2018
In this Project we will use Deep Neural Network to detect Diabetes.


## Importing Libraries

In [1]:
import sys
import pandas
import numpy
import sklearn
import keras



E:\Anaconda\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


### Getting data

In [2]:
import pandas as pd
import numpy as np

# import the uci pima indians diabetes dataset

df = pd.read_csv('diabetes.csv')

In [3]:
# Describe the dataset
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [4]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [5]:
df[df['Glucose'] == 0]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
75             1        0             48             20        0  24.7   
182            1        0             74             20       23  27.7   
342            1        0             68             35        0  32.0   
349            5        0             80             32        0  41.0   
502            6        0             68             41        0  39.0   

     DiabetesPedigreeFunction  Age  Outcome  
75                      0.140   22        0  
182                     0.299   21        0  
342                     0.389   22        0  
349                     0.346   37        1  
502                     0.727   41        1

In [6]:
# Preprocess the data, mark zero values as NaN and drop
columns = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

for col in columns:
    df[col].replace(0, np.NaN, inplace=True)
    
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  763.000000     733.000000     541.000000  394.000000   
mean      3.845052  121.686763      72.405184      29.153420  155.548223   
std       3.369578   30.535641      12.382158      10.476982  118.775855   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      64.000000      22.000000   76.250000   
50%       3.000000  117.000000      72.000000      29.000000  125.000000   
75%       6.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  757.000000                768.000000  768.000000  768.000000  
mean    32.457464                  0.471876   33.240885    0.348958  
std      6.924988                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.300000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [7]:
# Drop rows with missing values
df.dropna(inplace=True)

# summarize the number of rows and columns in df
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   392.000000  392.000000     392.000000     392.000000  392.000000   
mean      3.301020  122.627551      70.663265      29.145408  156.056122   
std       3.211424   30.860781      12.496092      10.516424  118.841690   
min       0.000000   56.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      62.000000      21.000000   76.750000   
50%       2.000000  119.000000      70.000000      29.000000  125.500000   
75%       5.000000  143.000000      78.000000      37.000000  190.000000   
max      17.000000  198.000000     110.000000      63.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  392.000000                392.000000  392.000000  392.000000  
mean    33.086224                  0.523046   30.864796    0.331633  
std      7.027659                  0.345488   10.200777    0.471401  
min     18.200000                  0.085000   21.000000    0.000000  
25%     28.400000                  0.269750   23.000000    0.000000  
50%     33.200000                  0.449500   27.000000    0.000000  
75%     37.100000                  0.687000   36.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [8]:
# split into input (X) and an output (Y)
X = df.iloc[:,0:8]
Y = df.iloc[:,8:]

In [9]:
X=X.values
Y=Y.values

In [10]:
print(X.shape)
print(Y.shape)
print(Y[:5])

(392, 8)
(392, 1)
[[0]
 [1]
 [1]
 [1]
 [1]]


In [11]:
# Normalize the data using sklearn StandardScaler
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X)

In [12]:
print(scaler)

StandardScaler(copy=True, with_mean=True, with_std=True)


In [13]:
# Transform and display the training data
X_standardized = scaler.transform(X)

data = pd.DataFrame(X_standardized)
data.describe()

0             1             2             3             4  \
count  3.920000e+02  3.920000e+02  3.920000e+02  3.920000e+02  3.920000e+02   
mean  -4.021726e-17  3.129583e-17 -4.641624e-16  1.042250e-16  6.485742e-17   
std    1.001278e+00  1.001278e+00  1.001278e+00  1.001278e+00  1.001278e+00   
min   -1.029213e+00 -2.161731e+00 -3.739001e+00 -2.108484e+00 -1.196867e+00   
25%   -7.174265e-01 -7.665958e-01 -6.941640e-01 -7.755315e-01 -6.681786e-01   
50%   -4.056403e-01 -1.176959e-01 -5.314565e-02 -1.384444e-02 -2.574448e-01   
75%    5.297185e-01  6.609841e-01  5.878727e-01  7.478426e-01  2.859877e-01   
max    4.271153e+00  2.445459e+00  3.151946e+00  3.223325e+00  5.812990e+00   

                  5             6             7  
count  3.920000e+02  3.920000e+02  3.920000e+02  
mean   1.543550e-16  3.880116e-17  1.028089e-16  
std    1.001278e+00  1.001278e+00  1.001278e+00  
min   -2.120941e+00 -1.269525e+00 -9.682991e-01  
25%   -6.676780e-01 -7.340909e-01 -7.719850e-01  
50%    1.621036e-02 -2.131475e-01 -3.793569e-01  
75%    5.718696e-01  4.751644e-01  5.040564e-01  
max    4.846172e+00  5.497667e+00  4.921123e+00

In [14]:
# import necessary sklearn and keras packages
from sklearn.model_selection import GridSearchCV, KFold
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.optimizers import Adam

In [15]:
# Start defining the model
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(8, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dense(4, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    
    # compile the model
    adam = Adam(lr = 0.01)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

model = create_model()
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 8)                 72        
_________________________________________________________________
dense_2 (Dense)              (None, 4)                 36        
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 5         
Total params: 113
Trainable params: 113
Non-trainable params: 0
_________________________________________________________________
None


In [16]:
# Do a grid search for the optimal batch size and number of epochs
# import necessary packages
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.optimizers import Adam

# Define a random seed
seed = 6
np.random.seed(seed)

# Start defining the model
def create_model():
    # create model
    model = Sequential()
    model.add(Dense(8, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dense(4, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    
    # compile the model
    adam = Adam(lr = 0.01)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

# create the model
model = KerasClassifier(build_fn = create_model, verbose = 1)

# define the grid search parameters
batch_size = [10, 20, 40]
epochs = [10, 50, 100]

# make a dictionary of the grid search parameters
param_grid = dict(batch_size=batch_size, epochs=epochs)

# build and fit the GridSearchCV
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = KFold(random_state=seed), verbose = 10)
grid_results = grid.fit(X_standardized, Y)

# summarize the results
print("Best: {0}, using {1}".format(grid_results.best_score_, grid_results.best_params_))
means = grid_results.cv_results_['mean_test_score']
stds = grid_results.cv_results_['std_test_score']
params = grid_results.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print('{0} ({1}) with: {2}'.format(mean, stdev, param))

Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] batch_size=10, epochs=10 ........................................
Epoch 1/10
261/261 [==============================] - 1s 4ms/step - loss: 0.6295 - acc: 0.6935
Epoch 2/10
261/261 [==============================] - 0s 268us/step - loss: 0.4853 - acc: 0.6973
Epoch 3/10
261/261 [==============================] - 0s 310us/step - loss: 0.4412 - acc: 0.6973
Epoch 4/10
261/261 [==============================] - 0s 291us/step - loss: 0.4247 - acc: 0.6973
Epoch 5/10
261/261 [==============================] - 0s 299us/step - loss: 0.4144 - acc: 0.8046
Epoch 6/10
261/261 [==============================] - 0s 318us/step - loss: 0.4019 - acc: 0.8276
Epoch 7/10
261/261 [==============================] - 0s 379us/step - loss: 0.3931 - acc: 0.8314
Epoch 8/10
261/261 [==============================] - 0s 280us/step - loss: 0.3902 - acc: 0.8199
Epoch 9/10
261/261 [==============================] - 0s 287us/step - loss: 0.3927 - acc: 0.827

[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    3.2s remaining:    0.0s


Epoch 1/10
261/261 [==============================] - 2s 7ms/step - loss: 0.6342 - acc: 0.6552
Epoch 2/10
261/261 [==============================] - 0s 352us/step - loss: 0.4960 - acc: 0.7471
Epoch 3/10
261/261 [==============================] - 0s 364us/step - loss: 0.4548 - acc: 0.7969
Epoch 4/10
261/261 [==============================] - 0s 364us/step - loss: 0.4422 - acc: 0.7893
Epoch 5/10
261/261 [==============================] - 0s 337us/step - loss: 0.4389 - acc: 0.7893
Epoch 6/10
261/261 [==============================] - 0s 295us/step - loss: 0.4339 - acc: 0.7969
Epoch 7/10
261/261 [==============================] - 0s 272us/step - loss: 0.4264 - acc: 0.8084
Epoch 8/10
261/261 [==============================] - 0s 326us/step - loss: 0.4184 - acc: 0.8161
Epoch 9/10
261/261 [==============================] - 0s 303us/step - loss: 0.4071 - acc: 0.8161
Epoch 10/10
261/261 [==============================] - 0s 123us/step
[CV]  batch_size=10, epochs=10, score=0.7633587745309786, to

[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    7.1s remaining:    0.0s


Epoch 1/10
262/262 [==============================] - 1s 4ms/step - loss: 0.6410 - acc: 0.6374
Epoch 2/10
262/262 [==============================] - 0s 275us/step - loss: 0.5272 - acc: 0.6527
Epoch 3/10
262/262 [==============================] - 0s 279us/step - loss: 0.5034 - acc: 0.7328
Epoch 4/10
262/262 [==============================] - 0s 290us/step - loss: 0.4995 - acc: 0.7519
Epoch 5/10
262/262 [==============================] - 0s 279us/step - loss: 0.4941 - acc: 0.7557
Epoch 6/10
262/262 [==============================] - 0s 313us/step - loss: 0.4885 - acc: 0.7595
Epoch 7/10
262/262 [==============================] - 0s 294us/step - loss: 0.4845 - acc: 0.7595
Epoch 8/10
262/262 [==============================] - 0s 294us/step - loss: 0.4798 - acc: 0.7672
Epoch 9/10
262/262 [==============================] - 0s 275us/step - loss: 0.4786 - acc: 0.7710
Epoch 10/10
262/262 [==============================] - 0s 122us/step
[CV]  batch_size=10, epochs=10, score=0.8153846034636865, to

[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:   10.3s remaining:    0.0s


Epoch 1/50
261/261 [==============================] - 1s 4ms/step - loss: 0.5895 - acc: 0.6820
Epoch 2/50
261/261 [==============================] - 0s 276us/step - loss: 0.4475 - acc: 0.6973
Epoch 3/50
261/261 [==============================] - 0s 318us/step - loss: 0.4235 - acc: 0.7318
Epoch 4/50
261/261 [==============================] - 0s 306us/step - loss: 0.4079 - acc: 0.8084
Epoch 5/50
261/261 [==============================] - 0s 272us/step - loss: 0.4029 - acc: 0.8238
Epoch 6/50
261/261 [==============================] - 0s 326us/step - loss: 0.3943 - acc: 0.8238
Epoch 7/50
261/261 [==============================] - 0s 291us/step - loss: 0.3809 - acc: 0.8314
Epoch 8/50
261/261 [==============================] - 0s 291us/step - loss: 0.3763 - acc: 0.8352
Epoch 9/50
261/261 [==============================] - 0s 283us/step - loss: 0.3840 - acc: 0.8199
Epoch 10/50
261/261 [==============================] - 0s 280us/step - loss: 0.3592 - acc: 0.8429
Epoch 11/50
261/261 [==========

[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:   16.5s remaining:    0.0s


Epoch 1/50
261/261 [==============================] - 1s 5ms/step - loss: 0.6597 - acc: 0.6858
Epoch 2/50
261/261 [==============================] - 0s 295us/step - loss: 0.5345 - acc: 0.7893
Epoch 3/50
261/261 [==============================] - 0s 322us/step - loss: 0.4534 - acc: 0.7969
Epoch 4/50
261/261 [==============================] - 0s 299us/step - loss: 0.4416 - acc: 0.8008
Epoch 5/50
261/261 [==============================] - 0s 299us/step - loss: 0.4277 - acc: 0.8084
Epoch 6/50
261/261 [==============================] - 0s 283us/step - loss: 0.4315 - acc: 0.8046
Epoch 7/50
261/261 [==============================] - 0s 299us/step - loss: 0.4168 - acc: 0.8046
Epoch 8/50
261/261 [==============================] - 0s 314us/step - loss: 0.4096 - acc: 0.8123
Epoch 9/50
261/261 [==============================] - 0s 307us/step - loss: 0.4075 - acc: 0.7969
Epoch 10/50
261/261 [==============================] - 0s 283us/step - loss: 0.4004 - acc: 0.8199
Epoch 11/50
261/261 [==========

[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:   23.1s remaining:    0.0s


Epoch 1/50
262/262 [==============================] - 1s 5ms/step - loss: 0.6051 - acc: 0.7061
Epoch 2/50
262/262 [==============================] - 0s 279us/step - loss: 0.4806 - acc: 0.7557
Epoch 3/50
262/262 [==============================] - 0s 290us/step - loss: 0.4735 - acc: 0.7557
Epoch 4/50
262/262 [==============================] - 0s 283us/step - loss: 0.4582 - acc: 0.7595
Epoch 5/50
262/262 [==============================] - 0s 301us/step - loss: 0.4517 - acc: 0.7710
Epoch 6/50
262/262 [==============================] - 0s 328us/step - loss: 0.4497 - acc: 0.7672
Epoch 7/50
262/262 [==============================] - 0s 286us/step - loss: 0.4427 - acc: 0.7863
Epoch 8/50
262/262 [==============================] - 0s 301us/step - loss: 0.4373 - acc: 0.7863
Epoch 9/50
262/262 [==============================] - 0s 294us/step - loss: 0.4381 - acc: 0.7863
Epoch 10/50
262/262 [==============================] - 0s 309us/step - loss: 0.4351 - acc: 0.7901
Epoch 11/50
262/262 [==========

[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:   30.5s remaining:    0.0s


Epoch 1/100
261/261 [==============================] - 2s 7ms/step - loss: 0.6346 - acc: 0.6667
Epoch 2/100
261/261 [==============================] - 0s 544us/step - loss: 0.4810 - acc: 0.6973
Epoch 3/100
261/261 [==============================] - 0s 494us/step - loss: 0.4362 - acc: 0.6973
Epoch 4/100
261/261 [==============================] - 0s 364us/step - loss: 0.4198 - acc: 0.7586
Epoch 5/100
261/261 [==============================] - 0s 337us/step - loss: 0.4088 - acc: 0.8238
Epoch 6/100
261/261 [==============================] - 0s 502us/step - loss: 0.3984 - acc: 0.8123
Epoch 7/100
261/261 [==============================] - 0s 418us/step - loss: 0.3901 - acc: 0.8276
Epoch 8/100
261/261 [==============================] - 0s 418us/step - loss: 0.3821 - acc: 0.8314
Epoch 9/100
261/261 [==============================] - 0s 456us/step - loss: 0.3739 - acc: 0.8429
Epoch 10/100
261/261 [==============================] - 0s 333us/step - loss: 0.3701 - acc: 0.8352
Epoch 11/100
261/261 

261/261 [==============================] - 0s 379us/step - loss: 0.2061 - acc: 0.9310
Epoch 84/100
261/261 [==============================] - 0s 318us/step - loss: 0.2069 - acc: 0.9119
Epoch 85/100
261/261 [==============================] - 0s 337us/step - loss: 0.2252 - acc: 0.9234
Epoch 86/100
261/261 [==============================] - 0s 379us/step - loss: 0.1929 - acc: 0.9310
Epoch 87/100
261/261 [==============================] - 0s 329us/step - loss: 0.1855 - acc: 0.9387
Epoch 88/100
261/261 [==============================] - 0s 383us/step - loss: 0.1967 - acc: 0.9272
Epoch 89/100
261/261 [==============================] - 0s 475us/step - loss: 0.1845 - acc: 0.9272
Epoch 90/100
261/261 [==============================] - 0s 326us/step - loss: 0.1875 - acc: 0.9310
Epoch 91/100
261/261 [==============================] - 0s 299us/step - loss: 0.1812 - acc: 0.9310
Epoch 92/100
261/261 [==============================] - 0s 291us/step - loss: 0.1784 - acc: 0.9349
Epoch 93/100
261/261 [=

[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:   44.1s remaining:    0.0s


Epoch 1/100
261/261 [==============================] - 3s 10ms/step - loss: 0.6835 - acc: 0.5977
Epoch 2/100
261/261 [==============================] - 0s 306us/step - loss: 0.6199 - acc: 0.7395
Epoch 3/100
261/261 [==============================] - 0s 303us/step - loss: 0.4716 - acc: 0.7816
Epoch 4/100
261/261 [==============================] - 0s 314us/step - loss: 0.4433 - acc: 0.7854
Epoch 5/100
261/261 [==============================] - 0s 291us/step - loss: 0.4275 - acc: 0.7931
Epoch 6/100
261/261 [==============================] - 0s 306us/step - loss: 0.4254 - acc: 0.8123
Epoch 7/100
261/261 [==============================] - 0s 287us/step - loss: 0.4254 - acc: 0.8046
Epoch 8/100
261/261 [==============================] - 0s 379us/step - loss: 0.4193 - acc: 0.8123
Epoch 9/100
261/261 [==============================] - 0s 356us/step - loss: 0.3997 - acc: 0.8238
Epoch 10/100
261/261 [==============================] - 0s 326us/step - loss: 0.3972 - acc: 0.8238
Epoch 11/100
261/261

261/261 [==============================] - 0s 303us/step - loss: 0.2659 - acc: 0.8927
Epoch 83/100
261/261 [==============================] - 0s 310us/step - loss: 0.2689 - acc: 0.8889
Epoch 84/100
261/261 [==============================] - 0s 529us/step - loss: 0.2626 - acc: 0.8927 0s - loss: 0.2632 - acc: 0.892
Epoch 85/100
261/261 [==============================] - 0s 977us/step - loss: 0.2571 - acc: 0.8966
Epoch 86/100
261/261 [==============================] - 0s 812us/step - loss: 0.2483 - acc: 0.8927
Epoch 87/100
261/261 [==============================] - 0s 632us/step - loss: 0.2960 - acc: 0.8736
Epoch 88/100
261/261 [==============================] - 0s 333us/step - loss: 0.2699 - acc: 0.8927
Epoch 89/100
261/261 [==============================] - 0s 299us/step - loss: 0.2599 - acc: 0.9004
Epoch 90/100
261/261 [==============================] - 0s 306us/step - loss: 0.2551 - acc: 0.8889
Epoch 91/100
261/261 [==============================] - 0s 318us/step - loss: 0.2544 - acc:

[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:   59.3s remaining:    0.0s


Epoch 1/100
262/262 [==============================] - 3s 11ms/step - loss: 0.6396 - acc: 0.6336
Epoch 2/100
262/262 [==============================] - 0s 408us/step - loss: 0.5347 - acc: 0.6527
Epoch 3/100
262/262 [==============================] - 0s 393us/step - loss: 0.5105 - acc: 0.6527
Epoch 4/100
262/262 [==============================] - 0s 305us/step - loss: 0.4948 - acc: 0.7328
Epoch 5/100
262/262 [==============================] - 0s 355us/step - loss: 0.4871 - acc: 0.7824
Epoch 6/100
262/262 [==============================] - 0s 397us/step - loss: 0.4787 - acc: 0.7634
Epoch 7/100
262/262 [==============================] - 0s 366us/step - loss: 0.4718 - acc: 0.7748
Epoch 8/100
262/262 [==============================] - 0s 531us/step - loss: 0.4670 - acc: 0.7901
Epoch 9/100
262/262 [==============================] - 0s 813us/step - loss: 0.4589 - acc: 0.7901
Epoch 10/100
262/262 [==============================] - 0s 824us/step - loss: 0.4512 - acc: 0.7977
Epoch 11/100
262/262

262/262 [==============================] - 0s 340us/step - loss: 0.3490 - acc: 0.8473
Epoch 84/100
262/262 [==============================] - 0s 363us/step - loss: 0.3444 - acc: 0.8473
Epoch 85/100
262/262 [==============================] - 0s 359us/step - loss: 0.3484 - acc: 0.8397
Epoch 86/100
262/262 [==============================] - 0s 366us/step - loss: 0.3490 - acc: 0.8550
Epoch 87/100
262/262 [==============================] - 0s 393us/step - loss: 0.3471 - acc: 0.8473
Epoch 88/100
262/262 [==============================] - 0s 511us/step - loss: 0.3485 - acc: 0.8511
Epoch 89/100
262/262 [==============================] - 0s 813us/step - loss: 0.3575 - acc: 0.8282
Epoch 90/100
262/262 [==============================] - 0s 809us/step - loss: 0.3594 - acc: 0.8359
Epoch 91/100
262/262 [==============================] - 0s 656us/step - loss: 0.3433 - acc: 0.8473
Epoch 92/100
262/262 [==============================] - 0s 324us/step - loss: 0.3394 - acc: 0.8550
Epoch 93/100
262/262 [=

[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:  1.3min remaining:    0.0s


Epoch 1/10
261/261 [==============================] - 2s 8ms/step - loss: 0.6725 - acc: 0.6743
Epoch 2/10
261/261 [==============================] - 0s 207us/step - loss: 0.5623 - acc: 0.7241
Epoch 3/10
261/261 [==============================] - 0s 161us/step - loss: 0.4395 - acc: 0.8008
Epoch 4/10
261/261 [==============================] - 0s 172us/step - loss: 0.4074 - acc: 0.8046
Epoch 5/10
261/261 [==============================] - 0s 180us/step - loss: 0.3915 - acc: 0.8199
Epoch 6/10
261/261 [==============================] - 0s 180us/step - loss: 0.3794 - acc: 0.8199
Epoch 7/10
261/261 [==============================] - 0s 360us/step - loss: 0.3735 - acc: 0.8314
Epoch 8/10
261/261 [==============================] - 0s 540us/step - loss: 0.3778 - acc: 0.8084
Epoch 9/10
261/261 [==============================] - 0s 372us/step - loss: 0.3778 - acc: 0.8123
Epoch 10/10
261/261 [==============================] - 0s 111us/step
[CV]  batch_size=20, epochs=10, score=0.7251908351446836, to

261/261 [==============================] - 0s 295us/step - loss: 0.3063 - acc: 0.8621
Epoch 47/50
261/261 [==============================] - 0s 341us/step - loss: 0.2976 - acc: 0.8582
Epoch 48/50
261/261 [==============================] - 0s 222us/step - loss: 0.2974 - acc: 0.8582
Epoch 49/50
261/261 [==============================] - 0s 345us/step - loss: 0.2905 - acc: 0.8659
Epoch 50/50
261/261 [==============================] - 0s 180us/step
[CV]  batch_size=20, epochs=50, score=0.7328244256609269, total=   8.7s
[CV] batch_size=20, epochs=50 ........................................
Epoch 1/50
261/261 [==============================] - 3s 11ms/step - loss: 0.6603 - acc: 0.6705
Epoch 2/50
261/261 [==============================] - 0s 299us/step - loss: 0.5549 - acc: 0.7663
Epoch 3/50
261/261 [==============================] - 0s 360us/step - loss: 0.4753 - acc: 0.7739
Epoch 4/50
261/261 [==============================] - 0s 414us/step - loss: 0.4450 - acc: 0.7854
Epoch 5/50
261/261 [=

262/262 [==============================] - 0s 214us/step - loss: 0.3861 - acc: 0.8321
Epoch 25/50
262/262 [==============================] - 0s 267us/step - loss: 0.3860 - acc: 0.8244
Epoch 26/50
262/262 [==============================] - 0s 202us/step - loss: 0.3807 - acc: 0.8321
Epoch 27/50
262/262 [==============================] - 0s 195us/step - loss: 0.4035 - acc: 0.8359
Epoch 28/50
262/262 [==============================] - 0s 309us/step - loss: 0.3814 - acc: 0.8206
Epoch 29/50
262/262 [==============================] - 0s 321us/step - loss: 0.3769 - acc: 0.8282
Epoch 30/50
262/262 [==============================] - 0s 214us/step - loss: 0.3738 - acc: 0.8244
Epoch 31/50
262/262 [==============================] - 0s 183us/step - loss: 0.3762 - acc: 0.8321
Epoch 32/50
262/262 [==============================] - 0s 214us/step - loss: 0.3693 - acc: 0.8321
Epoch 33/50
262/262 [==============================] - 0s 195us/step - loss: 0.3739 - acc: 0.8435
Epoch 34/50
262/262 [===========

261/261 [==============================] - 0s 207us/step - loss: 0.2646 - acc: 0.9042
Epoch 55/100
261/261 [==============================] - 0s 211us/step - loss: 0.3037 - acc: 0.8736
Epoch 56/100
261/261 [==============================] - 0s 203us/step - loss: 0.2860 - acc: 0.8774
Epoch 57/100
261/261 [==============================] - 0s 188us/step - loss: 0.2700 - acc: 0.8966
Epoch 58/100
261/261 [==============================] - 0s 215us/step - loss: 0.2689 - acc: 0.8927
Epoch 59/100
261/261 [==============================] - 0s 215us/step - loss: 0.2622 - acc: 0.9004
Epoch 60/100
261/261 [==============================] - 0s 192us/step - loss: 0.2596 - acc: 0.9042
Epoch 61/100
261/261 [==============================] - 0s 215us/step - loss: 0.2580 - acc: 0.9080
Epoch 62/100
261/261 [==============================] - 0s 211us/step - loss: 0.2579 - acc: 0.8966
Epoch 63/100
261/261 [==============================] - 0s 180us/step - loss: 0.2559 - acc: 0.9042
Epoch 64/100
261/261 [=

261/261 [==============================] - 0s 276us/step - loss: 0.3629 - acc: 0.8582
Epoch 35/100
261/261 [==============================] - 0s 287us/step - loss: 0.3557 - acc: 0.8736
Epoch 36/100
261/261 [==============================] - 0s 268us/step - loss: 0.3485 - acc: 0.8697
Epoch 37/100
261/261 [==============================] - 0s 268us/step - loss: 0.3408 - acc: 0.8736
Epoch 38/100
261/261 [==============================] - 0s 303us/step - loss: 0.3341 - acc: 0.8774
Epoch 39/100
261/261 [==============================] - 0s 337us/step - loss: 0.3343 - acc: 0.8736
Epoch 40/100
261/261 [==============================] - 0s 287us/step - loss: 0.3475 - acc: 0.8736
Epoch 41/100
261/261 [==============================] - 0s 272us/step - loss: 0.4322 - acc: 0.8046
Epoch 42/100
261/261 [==============================] - 0s 272us/step - loss: 0.4178 - acc: 0.8046
Epoch 43/100
261/261 [==============================] - 0s 268us/step - loss: 0.3566 - acc: 0.8621
Epoch 44/100
261/261 [=

262/262 [==============================] - 0s 187us/step - loss: 0.4259 - acc: 0.7863
Epoch 14/100
262/262 [==============================] - 0s 187us/step - loss: 0.4358 - acc: 0.7863
Epoch 15/100
262/262 [==============================] - 0s 183us/step - loss: 0.4310 - acc: 0.7863
Epoch 16/100
262/262 [==============================] - 0s 198us/step - loss: 0.4226 - acc: 0.8015
Epoch 17/100
262/262 [==============================] - 0s 172us/step - loss: 0.4187 - acc: 0.7939
Epoch 18/100
262/262 [==============================] - 0s 202us/step - loss: 0.4151 - acc: 0.8053
Epoch 19/100
262/262 [==============================] - 0s 179us/step - loss: 0.4156 - acc: 0.7977
Epoch 20/100
262/262 [==============================] - 0s 183us/step - loss: 0.4082 - acc: 0.8015
Epoch 21/100
262/262 [==============================] - 0s 195us/step - loss: 0.4052 - acc: 0.8168
Epoch 22/100
262/262 [==============================] - 0s 183us/step - loss: 0.4069 - acc: 0.8092
Epoch 23/100
262/262 [=

262/262 [==============================] - 0s 183us/step - loss: 0.3035 - acc: 0.8664
Epoch 96/100
262/262 [==============================] - 0s 202us/step - loss: 0.3034 - acc: 0.8664
Epoch 97/100
262/262 [==============================] - 0s 187us/step - loss: 0.2970 - acc: 0.8817
Epoch 98/100
262/262 [==============================] - 0s 175us/step - loss: 0.2972 - acc: 0.8817
Epoch 99/100
262/262 [==============================] - 0s 191us/step - loss: 0.2927 - acc: 0.8855
Epoch 100/100
262/262 [==============================] - 0s 118us/step
[CV]  batch_size=20, epochs=100, score=0.8384615457974948, total=   9.8s
[CV] batch_size=40, epochs=10 ........................................
Epoch 1/10
261/261 [==============================] - 3s 10ms/step - loss: 0.6723 - acc: 0.6820
Epoch 2/10
261/261 [==============================] - 0s 111us/step - loss: 0.6042 - acc: 0.6973
Epoch 3/10
261/261 [==============================] - 0s 115us/step - loss: 0.5141 - acc: 0.6973
Epoch 4/10
26

261/261 [==============================] - 0s 203us/step - loss: 0.3783 - acc: 0.8429
Epoch 39/50
261/261 [==============================] - 0s 126us/step - loss: 0.3762 - acc: 0.8352
Epoch 40/50
261/261 [==============================] - 0s 119us/step - loss: 0.3754 - acc: 0.8352
Epoch 41/50
261/261 [==============================] - 0s 123us/step - loss: 0.3760 - acc: 0.8429
Epoch 42/50
261/261 [==============================] - 0s 123us/step - loss: 0.3712 - acc: 0.8506
Epoch 43/50
261/261 [==============================] - 0s 119us/step - loss: 0.3679 - acc: 0.8352
Epoch 44/50
261/261 [==============================] - 0s 142us/step - loss: 0.3674 - acc: 0.8276
Epoch 45/50
261/261 [==============================] - 0s 134us/step - loss: 0.3630 - acc: 0.8467
Epoch 46/50
261/261 [==============================] - 0s 165us/step - loss: 0.3663 - acc: 0.8391
Epoch 47/50
261/261 [==============================] - 0s 333us/step - loss: 0.3603 - acc: 0.8391
Epoch 48/50
261/261 [===========

Epoch 16/50
262/262 [==============================] - 0s 134us/step - loss: 0.4422 - acc: 0.7863
Epoch 17/50
262/262 [==============================] - 0s 134us/step - loss: 0.4391 - acc: 0.7901
Epoch 18/50
262/262 [==============================] - 0s 114us/step - loss: 0.4422 - acc: 0.7710
Epoch 19/50
262/262 [==============================] - 0s 126us/step - loss: 0.4359 - acc: 0.7748
Epoch 20/50
262/262 [==============================] - 0s 118us/step - loss: 0.4307 - acc: 0.7901
Epoch 21/50
262/262 [==============================] - 0s 122us/step - loss: 0.4279 - acc: 0.7863
Epoch 22/50
262/262 [==============================] - 0s 122us/step - loss: 0.4256 - acc: 0.7824
Epoch 23/50
262/262 [==============================] - 0s 126us/step - loss: 0.4227 - acc: 0.7863
Epoch 24/50
262/262 [==============================] - 0s 122us/step - loss: 0.4225 - acc: 0.7901
Epoch 25/50
262/262 [==============================] - 0s 122us/step - loss: 0.4182 - acc: 0.7939
Epoch 26/50
262/262 

261/261 [==============================] - 0s 126us/step - loss: 0.3132 - acc: 0.8582
Epoch 46/100
261/261 [==============================] - 0s 126us/step - loss: 0.3093 - acc: 0.8544
Epoch 47/100
261/261 [==============================] - 0s 123us/step - loss: 0.3084 - acc: 0.8582
Epoch 48/100
261/261 [==============================] - 0s 126us/step - loss: 0.3086 - acc: 0.8621
Epoch 49/100
261/261 [==============================] - 0s 130us/step - loss: 0.3048 - acc: 0.8736
Epoch 50/100
261/261 [==============================] - 0s 145us/step - loss: 0.3004 - acc: 0.8697
Epoch 51/100
261/261 [==============================] - 0s 123us/step - loss: 0.3009 - acc: 0.8697
Epoch 52/100
261/261 [==============================] - 0s 123us/step - loss: 0.3005 - acc: 0.8774
Epoch 53/100
261/261 [==============================] - 0s 123us/step - loss: 0.2971 - acc: 0.8774
Epoch 54/100
261/261 [==============================] - 0s 123us/step - loss: 0.2934 - acc: 0.8774
Epoch 55/100
261/261 [=

Epoch 25/100
261/261 [==============================] - 0s 138us/step - loss: 0.3768 - acc: 0.8429
Epoch 26/100
261/261 [==============================] - 0s 130us/step - loss: 0.3742 - acc: 0.8429
Epoch 27/100
261/261 [==============================] - 0s 138us/step - loss: 0.3717 - acc: 0.8391
Epoch 28/100
261/261 [==============================] - 0s 126us/step - loss: 0.3724 - acc: 0.8506
Epoch 29/100
261/261 [==============================] - 0s 138us/step - loss: 0.3671 - acc: 0.8467
Epoch 30/100
261/261 [==============================] - 0s 138us/step - loss: 0.3675 - acc: 0.8391
Epoch 31/100
261/261 [==============================] - 0s 138us/step - loss: 0.3631 - acc: 0.8506
Epoch 32/100
261/261 [==============================] - 0s 146us/step - loss: 0.3627 - acc: 0.8467
Epoch 33/100
261/261 [==============================] - 0s 142us/step - loss: 0.3609 - acc: 0.8582
Epoch 34/100
261/261 [==============================] - 0s 176us/step - loss: 0.3592 - acc: 0.8506
Epoch 35/1

262/262 [==============================] - 0s 130us/step - loss: 0.5163 - acc: 0.7519
Epoch 5/100
262/262 [==============================] - 0s 118us/step - loss: 0.4835 - acc: 0.7557
Epoch 6/100
262/262 [==============================] - 0s 137us/step - loss: 0.4721 - acc: 0.7519
Epoch 7/100
262/262 [==============================] - 0s 134us/step - loss: 0.4729 - acc: 0.7519
Epoch 8/100
262/262 [==============================] - 0s 134us/step - loss: 0.4642 - acc: 0.7710
Epoch 9/100
262/262 [==============================] - 0s 134us/step - loss: 0.4631 - acc: 0.7519
Epoch 10/100
262/262 [==============================] - 0s 134us/step - loss: 0.4608 - acc: 0.7672
Epoch 11/100
262/262 [==============================] - 0s 145us/step - loss: 0.4541 - acc: 0.7519
Epoch 12/100
262/262 [==============================] - 0s 141us/step - loss: 0.4542 - acc: 0.7710
Epoch 13/100
262/262 [==============================] - 0s 149us/step - loss: 0.4533 - acc: 0.7710
Epoch 14/100
262/262 [======

Epoch 87/100
262/262 [==============================] - 0s 141us/step - loss: 0.3182 - acc: 0.8702
Epoch 88/100
262/262 [==============================] - 0s 198us/step - loss: 0.3185 - acc: 0.8664
Epoch 89/100
262/262 [==============================] - 0s 149us/step - loss: 0.3124 - acc: 0.8740
Epoch 90/100
262/262 [==============================] - 0s 160us/step - loss: 0.3088 - acc: 0.8664
Epoch 91/100
262/262 [==============================] - 0s 214us/step - loss: 0.3065 - acc: 0.8740
Epoch 92/100
262/262 [==============================] - 0s 164us/step - loss: 0.3074 - acc: 0.8702
Epoch 93/100
262/262 [==============================] - 0s 145us/step - loss: 0.3051 - acc: 0.8779
Epoch 94/100
262/262 [==============================] - 0s 137us/step - loss: 0.3041 - acc: 0.8817
Epoch 95/100
262/262 [==============================] - 0s 145us/step - loss: 0.3036 - acc: 0.8817
Epoch 96/100
262/262 [==============================] - 0s 141us/step - loss: 0.2974 - acc: 0.8817
Epoch 97/1

[Parallel(n_jobs=1)]: Done  27 out of  27 | elapsed:  3.9min finished


Epoch 1/50
392/392 [==============================] - 3s 9ms/step - loss: 0.6786 - acc: 0.6811
Epoch 2/50
392/392 [==============================] - 0s 102us/step - loss: 0.6199 - acc: 0.7704
Epoch 3/50
392/392 [==============================] - 0s 102us/step - loss: 0.5326 - acc: 0.7730
Epoch 4/50
392/392 [==============================] - 0s 107us/step - loss: 0.4640 - acc: 0.7755
Epoch 5/50
392/392 [==============================] - 0s 110us/step - loss: 0.4447 - acc: 0.7755
Epoch 6/50
392/392 [==============================] - 0s 107us/step - loss: 0.4389 - acc: 0.7781
Epoch 7/50
392/392 [==============================] - 0s 115us/step - loss: 0.4342 - acc: 0.7857
Epoch 8/50
392/392 [==============================] - 0s 99us/step - loss: 0.4317 - acc: 0.7908
Epoch 9/50
392/392 [==============================] - 0s 107us/step - loss: 0.4273 - acc: 0.7985
Epoch 10/50
392/392 [==============================] - 0s 115us/step - loss: 0.4285 - acc: 0.8010
Epoch 11/50
392/392 [===========

In [18]:
# Do a grid search for learning rate and dropout rate
# import necessary packages
from keras.layers import Dropout

# Define a random seed
seed = 6
np.random.seed(seed)

# Start defining the model
def create_model(learn_rate, dropout_rate):
    # create model
    model = Sequential()
    model.add(Dense(8, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dropout(dropout_rate))
    model.add(Dense(4, input_dim = 8, kernel_initializer='normal', activation='relu'))
    model.add(Dropout(dropout_rate))
    model.add(Dense(1, activation='sigmoid'))
    
    # compile the model
    adam = Adam(lr = learn_rate)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

# create the model
model = KerasClassifier(build_fn = create_model, epochs = 100, batch_size = 20, verbose = 0)

# define the grid search parameters
learn_rate = [0.001, 0.01, 0.1]
dropout_rate = [0.0, 0.1, 0.2]

# make a dictionary of the grid search parameters
param_grid = dict(learn_rate=learn_rate, dropout_rate=dropout_rate)

# build and fit the GridSearchCV
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = KFold(random_state=seed), verbose = 10)
grid_results = grid.fit(X_standardized, Y)

# summarize the results
print("Best: {0}, using {1}".format(grid_results.best_score_, grid_results.best_params_))
means = grid_results.cv_results_['mean_test_score']
stds = grid_results.cv_results_['std_test_score']
params = grid_results.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print('{0} ({1}) with: {2}'.format(mean, stdev, param))

Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] dropout_rate=0.0, learn_rate=0.001 ..............................
[CV]  dropout_rate=0.0, learn_rate=0.001, score=0.74809161397337, total=  19.9s
[CV] dropout_rate=0.0, learn_rate=0.001 ..............................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   20.0s remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.001, score=0.7709923705071894, total=  16.8s
[CV] dropout_rate=0.0, learn_rate=0.001 ..............................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:   36.9s remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.001, score=0.8538461556801429, total=  16.7s
[CV] dropout_rate=0.0, learn_rate=0.01 ...............................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:   53.7s remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.01, score=0.7480916130633755, total=  17.0s
[CV] dropout_rate=0.0, learn_rate=0.01 ...............................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:  1.2min remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.01, score=0.7709923705071894, total=  17.5s
[CV] dropout_rate=0.0, learn_rate=0.01 ...............................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:  1.5min remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.01, score=0.8076923122772803, total=  17.5s
[CV] dropout_rate=0.0, learn_rate=0.1 ................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:  1.8min remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.1, score=0.7328244247509323, total=  17.8s
[CV] dropout_rate=0.0, learn_rate=0.1 ................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:  2.1min remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.1, score=0.7099236750420723, total=  20.6s
[CV] dropout_rate=0.0, learn_rate=0.1 ................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:  2.4min remaining:    0.0s


[CV]  dropout_rate=0.0, learn_rate=0.1, score=0.830769236271198, total=  27.8s
[CV] dropout_rate=0.1, learn_rate=0.001 ..............................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:  2.9min remaining:    0.0s


[CV]  dropout_rate=0.1, learn_rate=0.001, score=0.725190845154624, total=  26.8s
[CV] dropout_rate=0.1, learn_rate=0.001 ..............................
[CV]  dropout_rate=0.1, learn_rate=0.001, score=0.7557252003946378, total=  25.6s
[CV] dropout_rate=0.1, learn_rate=0.001 ..............................
[CV]  dropout_rate=0.1, learn_rate=0.001, score=0.8384615366275494, total=  22.9s
[CV] dropout_rate=0.1, learn_rate=0.01 ...............................
[CV]  dropout_rate=0.1, learn_rate=0.01, score=0.7480916039634297, total=  21.9s
[CV] dropout_rate=0.1, learn_rate=0.01 ...............................
[CV]  dropout_rate=0.1, learn_rate=0.01, score=0.7786259501034977, total=  21.7s
[CV] dropout_rate=0.1, learn_rate=0.01 ...............................
[CV]  dropout_rate=0.1, learn_rate=0.01, score=0.8230769267449012, total=  21.8s
[CV] dropout_rate=0.1, learn_rate=0.1 ................................
[CV]  dropout_rate=0.1, learn_rate=0.1, score=0.7709923727821758, total=  22.6s
[CV] d

[Parallel(n_jobs=1)]: Done  27 out of  27 | elapsed: 11.0min finished


Best: 0.7908163321565609, using {'dropout_rate': 0.0, 'learn_rate': 0.001}
0.7908163321565609 (0.04537451427336758) with: {'dropout_rate': 0.0, 'learn_rate': 0.001}
0.7755102103157919 (0.02452593043414963) with: {'dropout_rate': 0.0, 'learn_rate': 0.01}
0.7576530657860697 (0.05234704460030959) with: {'dropout_rate': 0.0, 'learn_rate': 0.1}
0.7729591880829967 (0.04779843308896438) with: {'dropout_rate': 0.1, 'learn_rate': 0.001}
0.7831632654581752 (0.03076129219834126) with: {'dropout_rate': 0.1, 'learn_rate': 0.01}
0.7525510254259012 (0.01656379453153207) with: {'dropout_rate': 0.1, 'learn_rate': 0.1}
0.7806122511321184 (0.043633878203176005) with: {'dropout_rate': 0.2, 'learn_rate': 0.001}
0.7780612279870072 (0.015765620733424204) with: {'dropout_rate': 0.2, 'learn_rate': 0.01}
0.6964285645862015 (0.006728724114385511) with: {'dropout_rate': 0.2, 'learn_rate': 0.1}


In [20]:
# Do a grid search to optimize kernel initialization and activation functions
# import necessary packages

# Define a random seed
seed = 6
np.random.seed(seed)

# Start defining the model
def create_model(activation, init):
    # create model
    model = Sequential()
    model.add(Dense(8, input_dim = 8, kernel_initializer= init, activation= activation))
    model.add(Dense(4, input_dim = 8, kernel_initializer= init, activation= activation))
    model.add(Dense(1, activation='sigmoid'))
    
    # compile the model
    adam = Adam(lr = 0.001)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

# create the model
model = KerasClassifier(build_fn = create_model, epochs = 100, batch_size = 20, verbose = 0)

# define the grid search parameters
activation = ['softmax', 'relu', 'tanh', 'linear']
init = ['uniform', 'normal', 'zero']

# make a dictionary of the grid search parameters
param_grid = dict(activation = activation, init = init)

# build and fit the GridSearchCV
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = KFold(random_state=seed), verbose = 10)
grid_results = grid.fit(X_standardized, Y)

# summarize the results
print("Best: {0}, using {1}".format(grid_results.best_score_, grid_results.best_params_))
means = grid_results.cv_results_['mean_test_score']
stds = grid_results.cv_results_['std_test_score']
params = grid_results.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print('{0} ({1}) with: {2}'.format(mean, stdev, param))

Fitting 3 folds for each of 12 candidates, totalling 36 fits
[CV] activation=softmax, init=uniform ................................
[CV]  activation=softmax, init=uniform, score=0.7557252035796187, total= 2.4min
[CV] activation=softmax, init=uniform ................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  2.4min remaining:    0.0s


[CV]  activation=softmax, init=uniform, score=0.7557252003946378, total= 1.2min
[CV] activation=softmax, init=uniform ................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  3.6min remaining:    0.0s


[CV]  activation=softmax, init=uniform, score=0.8153846218035772, total=  48.3s
[CV] activation=softmax, init=normal .................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  4.4min remaining:    0.0s


[CV]  activation=softmax, init=normal, score=0.6106870242657553, total=  49.3s
[CV] activation=softmax, init=normal .................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:  5.2min remaining:    0.0s


[CV]  activation=softmax, init=normal, score=0.7557252003946378, total=  52.5s
[CV] activation=softmax, init=normal .................................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:  6.1min remaining:    0.0s


[CV]  activation=softmax, init=normal, score=0.8230769267449012, total=  44.9s
[CV] activation=softmax, init=zero ...................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:  6.9min remaining:    0.0s


[CV]  activation=softmax, init=zero, score=0.6106870242657553, total=  50.7s
[CV] activation=softmax, init=zero ...................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:  7.7min remaining:    0.0s


[CV]  activation=softmax, init=zero, score=0.7557252003946378, total=  45.6s
[CV] activation=softmax, init=zero ...................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:  8.5min remaining:    0.0s


[CV]  activation=softmax, init=zero, score=0.8230769313298739, total=  52.1s
[CV] activation=relu, init=uniform ...................................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:  9.4min remaining:    0.0s


[CV]  activation=relu, init=uniform, score=0.7251908542545698, total=  46.3s
[CV] activation=relu, init=uniform ...................................
[CV]  activation=relu, init=uniform, score=0.755725191294692, total=  43.4s
[CV] activation=relu, init=uniform ...................................
[CV]  activation=relu, init=uniform, score=0.8230769221599286, total=  46.1s
[CV] activation=relu, init=normal ....................................
[CV]  activation=relu, init=normal, score=0.7175572455384349, total= 4.1min
[CV] activation=relu, init=normal ....................................
[CV]  activation=relu, init=normal, score=0.7709923705071894, total= 1.1min
[CV] activation=relu, init=normal ....................................
[CV]  activation=relu, init=normal, score=0.8230769267449012, total= 1.2min
[CV] activation=relu, init=zero ......................................
[CV]  activation=relu, init=zero, score=0.6106870242657553, total= 1.4min
[CV] activation=relu, init=zero ..........

[Parallel(n_jobs=1)]: Done  36 out of  36 | elapsed: 37.8min finished


Best: 0.7933673531729348, using {'activation': 'tanh', 'init': 'normal'}
0.7755102136609505 (0.028087641954935023) with: {'activation': 'softmax', 'init': 'uniform'}
0.7295918416003792 (0.08860773018104312) with: {'activation': 'softmax', 'init': 'normal'}
0.7295918431209059 (0.0886077317852662) with: {'activation': 'softmax', 'init': 'zero'}
0.7678571475707755 (0.04085046948442003) with: {'activation': 'relu', 'init': 'uniform'}
0.7704081637214641 (0.04305242085893215) with: {'activation': 'relu', 'init': 'normal'}
0.6683673458744068 (0.040922317011154695) with: {'activation': 'relu', 'init': 'zero'}
0.7831632721484924 (0.02879958859349298) with: {'activation': 'tanh', 'init': 'uniform'}
0.7933673531729348 (0.032371719468446684) with: {'activation': 'tanh', 'init': 'normal'}
0.6683673458744068 (0.040922317011154695) with: {'activation': 'tanh', 'init': 'zero'}
0.7933673531729348 (0.037313655933021314) with: {'activation': 'linear', 'init': 'uniform'}
0.7908163291155076 (0.033706165933

In [22]:
# Do a grid search to find the optimal number of neurons in each hidden layer
# import necessary packages

# Define a random seed
seed = 6
np.random.seed(seed)

# Start defining the model
def create_model(neuron1, neuron2):
    # create model
    model = Sequential()
    model.add(Dense(neuron1, input_dim = 8, kernel_initializer= 'uniform', activation= 'linear'))
    model.add(Dense(neuron2, input_dim = neuron1, kernel_initializer= 'uniform', activation= 'linear'))
    model.add(Dense(1, activation='sigmoid'))
    
    # compile the model
    adam = Adam(lr = 0.001)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

# create the model
model = KerasClassifier(build_fn = create_model, epochs = 100, batch_size = 20, verbose = 0)

# define the grid search parameters
neuron1 = [4, 8, 16]
neuron2 = [2, 4, 8]

# make a dictionary of the grid search parameters
param_grid = dict(neuron1 = neuron1, neuron2 = neuron2)

# build and fit the GridSearchCV
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = KFold(random_state=seed), refit = True, verbose = 10)
grid_results = grid.fit(X_standardized, Y)

# summarize the results
print("Best: {0}, using {1}".format(grid_results.best_score_, grid_results.best_params_))
means = grid_results.cv_results_['mean_test_score']
stds = grid_results.cv_results_['std_test_score']
params = grid_results.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print('{0} ({1}) with: {2}'.format(mean, stdev, param))

Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] neuron1=4, neuron2=2 ............................................
[CV] ... neuron1=4, neuron2=2, score=0.7709923645922245, total= 1.3min
[CV] neuron1=4, neuron2=2 ............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:  1.3min remaining:    0.0s


[CV] ... neuron1=4, neuron2=2, score=0.7633587900008864, total= 1.1min
[CV] neuron1=4, neuron2=2 ............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  2.5min remaining:    0.0s


[CV] ... neuron1=4, neuron2=2, score=0.8230769267449012, total= 1.1min
[CV] neuron1=4, neuron2=4 ............................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  3.6min remaining:    0.0s


[CV] ... neuron1=4, neuron2=4, score=0.7709923645922245, total= 1.2min
[CV] neuron1=4, neuron2=4 ............................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:  4.8min remaining:    0.0s


[CV] ... neuron1=4, neuron2=4, score=0.7786259692133838, total= 1.2min
[CV] neuron1=4, neuron2=4 ............................................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:  6.0min remaining:    0.0s


[CV] ... neuron1=4, neuron2=4, score=0.8153846218035772, total= 1.2min
[CV] neuron1=4, neuron2=8 ............................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:  7.2min remaining:    0.0s


[CV] ... neuron1=4, neuron2=8, score=0.7633587749859759, total= 1.2min
[CV] neuron1=4, neuron2=8 ............................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:  8.3min remaining:    0.0s


[CV] ... neuron1=4, neuron2=8, score=0.7633587900008864, total= 1.2min
[CV] neuron1=4, neuron2=8 ............................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:  9.5min remaining:    0.0s


[CV] .... neuron1=4, neuron2=8, score=0.830769236271198, total= 1.2min
[CV] neuron1=8, neuron2=2 ............................................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed: 10.7min remaining:    0.0s


[CV] ... neuron1=8, neuron2=2, score=0.7633587840859216, total= 1.1min
[CV] neuron1=8, neuron2=2 ............................................
[CV] ... neuron1=8, neuron2=2, score=0.7633587900008864, total= 1.1min
[CV] neuron1=8, neuron2=2 ............................................
[CV] ... neuron1=8, neuron2=2, score=0.8384615457974948, total= 1.1min
[CV] neuron1=8, neuron2=4 ............................................
[CV] ... neuron1=8, neuron2=4, score=0.7633587749859759, total= 1.1min
[CV] neuron1=8, neuron2=4 ............................................
[CV] ... neuron1=8, neuron2=4, score=0.7633587900008864, total= 1.3min
[CV] neuron1=8, neuron2=4 ............................................
[CV] ... neuron1=8, neuron2=4, score=0.8384615457974948, total= 1.3min
[CV] neuron1=8, neuron2=8 ............................................
[CV] ... neuron1=8, neuron2=8, score=0.7633587840859216, total= 1.1min
[CV] neuron1=8, neuron2=8 ............................................
[CV] .

[Parallel(n_jobs=1)]: Done  27 out of  27 | elapsed: 32.2min finished


Best: 0.7908163351976142, using {'neuron1': 16, 'neuron2': 2}
0.785714290123813 (0.02650267677503863) with: {'neuron1': 4, 'neuron2': 2}
0.7882653126607135 (0.019356087898682556) with: {'neuron1': 4, 'neuron2': 4}
0.785714290123813 (0.03173682706863349) with: {'neuron1': 4, 'neuron2': 8}
0.7882653141812402 (0.03535836258888925) with: {'neuron1': 8, 'neuron2': 2}
0.7882653111401869 (0.03535836473101593) with: {'neuron1': 8, 'neuron2': 4}
0.7857142931648663 (0.031736824926506764) with: {'neuron1': 8, 'neuron2': 8}
0.7908163351976142 (0.03897990025127174) with: {'neuron1': 16, 'neuron2': 2}
0.7908163306360342 (0.03370616199600475) with: {'neuron1': 16, 'neuron2': 4}
0.7857142931648663 (0.031736824926506764) with: {'neuron1': 16, 'neuron2': 8}


In [23]:
# generate predictions with optimal hyperparameters
y_pred = grid.predict(X_standardized)

In [24]:
print(y_pred.shape)

(392, 1)


In [25]:
print(y_pred[:5])

[[0]
 [1]
 [0]
 [1]
 [1]]


In [27]:
# Generate a classification report
from sklearn.metrics import classification_report, accuracy_score,confusion_matrix

print(accuracy_score(Y, y_pred))
print(classification_report(Y, y_pred))

0.7806122448979592
             precision    recall  f1-score   support

          0       0.81      0.89      0.84       262
          1       0.71      0.57      0.63       130

avg / total       0.77      0.78      0.77       392



In [28]:
print(confusion_matrix(Y, y_pred))

[[232  30]
 [ 56  74]]


<b>Accuracy is 78%</b>

You all are invited to contribute and improve the performance.

# Thank You